evm/rpc/ipc.py

```diff
@@ -0,0 +1,127 @@
+import asyncio
+import json
+import logging
+import os
+
+from cytoolz import curry
+from eth_utils import decode_hex
+
+from evm import MainnetTesterChain
+from evm.db.backends.memory import MemoryDB
+from evm.db.chain import BaseChainDB
+from evm.rpc.main import RPCServer
+
+
+MAXIMUM_REQUEST_BYTES = 10000
+
+
+@curry
+async def connection_handler(execute_rpc, reader, writer):
+    '''
+    Catch fatal errors, log them, and close the connection
+    '''
+    logger = logging.getLogger('evm.rpc.ipc')
+    try:
+        await connection_loop(execute_rpc, reader, writer, logger)
+    except (ConnectionResetError, asyncio.IncompleteReadError):
+        logger.debug("Client closed connection")
+    except Exception:
+        logger.exception("Unrecognized exception while handling requests")
+    writer.close()
+
+
+async def connection_loop(execute_rpc, reader, writer, logger):
+    raw_request = ''
+    while True:
+        request_bytes = b''
+        try:
+            request_bytes = await reader.readuntil(b'}')
+        except asyncio.LimitOverrunError as e:
+            logger.info("Client request was too long. Erasing buffer and restarting...")
+            request_bytes = await reader.read(e.consumed)
+            await write_error(writer, "reached limit: %d bytes, starting with '%s'" % (
+                e.consumed,
+                request_bytes[:20],
+            ))
+            continue
+
+        raw_request += request_bytes.decode()
+
+        bad_prefix, raw_request = strip_non_json_prefix(raw_request)
+        if bad_prefix:
+            logger.info("Client started request with non json data: %r", bad_prefix)
+            await write_error(writer, 'Cannot parse json: ' + bad_prefix)
+
+        try:
+            request = json.loads(raw_request)
+        except json.JSONDecodeError:
+            # invalid json request, keep reading data until a valid json is formed
+            logger.debug("Invalid JSON, waiting for rest of message: %r", raw_request)
+            continue
+
+        # reset the buffer for the next message
+        raw_request = ''
+
+        if not request:
+            logger.debug("Client sent empty request")
+            continue
+
+        try:
+            result = execute_rpc(request)
+        except Exception as e:
+            logger.exception("Unrecognized exception while executing RPC")
+            await write_error(writer, "unknown failure: " + str(e))
+        else:
+            writer.write(result.encode())
+
+        await writer.drain()
+
+
+def strip_non_json_prefix(raw_request):
+    if raw_request and raw_request[0] != '{':
+        prefix, bracket, rest = raw_request.partition('{')
+        return prefix.strip(), bracket + rest
+    else:
+        return '', raw_request
+
+
+async def write_error(writer, message):
+    json_error = '{"error": "%s"}\n' % message
```

Comment:
I think constructing json using string formatting is asking for problems since the message could have characters that need to be escaped.  Can we use a legit JSON encoder here?